content/automate/ManageIQ/Transformation/TransformationHosts/Common.class/__methods__/utils.rb

```diff
@@ -33,49 +30,77 @@ def self.transformation_hosts(ems, factory_config)
+            def self.get_runners_count_by_ems(ems, factory_config, handle = $evm)
+              transformation_hosts(ems, factory_config, handle).inject(0) { |sum, thost| sum + thost[:runners][:current] }
             end
 
-            def self.ems_max_runners(ems, factory_config, max_runners = DEFAULT_EMS_MAX_RUNNERS, handle = $evm)
+            def self.ems_max_runners(ems, factory_config, max_runners = DEFAULT_EMS_MAX_RUNNERS)
               if ems.custom_get('Max Transformation Runners')
-                handle.log(:info, "Using max transformation runners from EMS custom attribute: #{ems.custom_get('Max Transformation Runners')}")
                 ems.custom_get('Max Transformation Runners').to_i
               elsif factory_config['ems_max_runners']
-                handle.log(:info, "Using max transformation runners from factory config: #{factory_config['ems_max_runners']}")
                 factory_config['ems_max_runners'].to_i
               else
-                handle.log(:info, "Using default max transformation runners: #{max_runners}")
                 max_runners
               end
             end
 
             def self.get_transformation_host(task, factory_config, handle = $evm)
               ems = handle.vmdb(:ext_management_system).find_by(:id => task.get_option(:destination_ems_id))
-              ems_cur_runners = get_runners_count_by_ems(ems, factory_config)
-              transformation_host_hash = ems_cur_runners < ems_max_runners(ems, factory_config) ? eligible_transformation_hosts(ems, factory_config).first : {}
-              return transformation_host_hash[:type], transformation_host_hash[:host], transformation_host_hash[:transformation_method]
+              ems_cur_runners = get_runners_count_by_ems(ems, factory_config, handle)
+              return unless ems_cur_runners < ems_max_runners(ems, factory_config)
+              thosts = eligible_transformation_hosts(ems, factory_config, handle)
+              return if thosts.size.zero?
+              [thosts.first[:type], thosts.first[:host], thosts.first[:transformation_method]]
             end
 
             def self.virtv2vwrapper_options(task)
               send("virtv2vwrapper_options_#{task.get_option(:transformation_type)}_#{task.get_option(:transformation_method)}", task)
             end
 
-            def self.remote_command(task, transformation_host, command, stdin = nil, run_as = nil)
-              "ManageIQ::Automate::Transformation::TransformationHosts::#{task.get_option(:transformation_host_type)}::Utils".constantize.remote_command(transformation_host, command, stdin, run_as)
+            def self.vmware_uri_vddk(vm)
+              URI::Generic.build(
+                :scheme   => 'esx',
+                :userinfo => CGI.escape(vm.host.authentication_userid),
+                :host     => vm.host.ipaddress,
+                :path     => '/',
+                :query    => { :no_verify => 1 }.to_query
```

Comment:
What is the to_query method on hash?

Reply:
It converts this string to `?no_verify=1`, which convenient to escape input.